datastore/src/main/java/io/spine/server/storage/datastore/DsRecordStorage.java

```diff
@@ -212,40 +213,92 @@ public EntityRecord apply(@Nullable Entity input) {
         }
-        // Only column query
+        return result;
+    }
+
+    /**
+     * Performs a query by IDs and entity columns.
+     *
+     * <p>The by-IDs query is performed on Datastore, and the by-columns filtering is done in
+     * memory.
+     *
+     * @param acceptableIds the IDs to search by
+     * @param params        the additional query parameters
+     * @param fieldMask     the {@code FieldMask} to apply to all the retrieved entity states
+     * @return an iterator over the resulting entity records
+     */
+    private Iterator<EntityRecord> queryByIdsAndColumns(Collection<I> acceptableIds,
+                                                        QueryParameters params,
+                                                        FieldMask fieldMask) {
+        final Predicate<Entity> inMemPredicate;
+        if (params.iterator().hasNext()) { // IDs and columns query
+            inMemPredicate = buildMemoryPredicate(params);
+        } else { // Only IDs query
+            inMemPredicate = alwaysTrue();
+        }
+        final Iterator<EntityRecord> records = lookup(acceptableIds, fieldMask, inMemPredicate);
+        return records;
+    }
+
+    /**
+     * Performs a query by entity columns.
+     *
+     * <p>The query is performed on Datastore. A single call to this method may turn into a few API
+     * calls. See {@link DsFilters} for details.
+     *
+     * @param params    the by-column query parameters
+     * @param fieldMask the {@code FieldMask} to apply to all the retrieved entity states
+     * @return an iterator over the resulting entity records
+     */
+    private Iterator<EntityRecord> queryByColumnsOnly(QueryParameters params,
+                                                      FieldMask fieldMask) {
         final StructuredQuery.Builder<Entity> datastoreQuery = Query.newEntityQueryBuilder()
                                                                     .setKind(getKind().getValue());
         final Collection<Filter> filters = buildColumnFilters(params);
-        final Collection<StructuredQuery<Entity>> queries = transform(
-                filters,
-                new Function<Filter, StructuredQuery<Entity>>() {
-                    @Override
-                    public StructuredQuery<Entity> apply(@Nullable Filter input) {
-                        checkNotNull(input);
-                        return datastoreQuery.setFilter(input)
-                                             .build();
-                    }
-                }
-        );
+        final FilterToQuery function = new FilterToQuery(datastoreQuery);
+        final Collection<StructuredQuery<Entity>> queries = transform(filters, function);
+        final Iterator<EntityRecord> result = queryAndMerge(queries, fieldMask);
+        return result;
+    }
+
+    /**
+     * Performs the given Datastore {@linkplain StructuredQuery queries} and combines results into
+     * a single lazy iterator.
```

Comment:
Please explain the lazy nature in the following paragraphs.

Reply:
Done.